File: app/src/main/java/wangdaye/com/geometricweather/weather/services/CaiYunWeatherService.java
package wangdaye.com.geometricweather.weather.services;

import android.content.Context;

import androidx.annotation.NonNull;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.disposables.CompositeDisposable;
import wangdaye.com.geometricweather.common.basic.models.ChineseCity;
import wangdaye.com.geometricweather.common.basic.models.Location;
import wangdaye.com.geometricweather.common.rxjava.BaseObserver;
import wangdaye.com.geometricweather.common.rxjava.ObserverContainer;
import wangdaye.com.geometricweather.common.rxjava.SchedulerTransformer;
import wangdaye.com.geometricweather.weather.apis.CNWeatherApi;
import wangdaye.com.geometricweather.weather.apis.CaiYunApi;
import wangdaye.com.geometricweather.weather.converters.CaiyunResultConverter;
import wangdaye.com.geometricweather.weather.json.caiyun.CaiYunForecastResult;
import wangdaye.com.geometricweather.weather.json.caiyun.CaiYunMainlyResult;

/**
 * CaiYun weather service.
 * */

public class CaiYunWeatherService extends CNWeatherService {

    private final CaiYunApi mApi;
    private final CompositeDisposable mCompositeDisposable;

    @Inject
    public CaiYunWeatherService(CaiYunApi cyApi, CNWeatherApi cnApi, CompositeDisposable disposable) {
        super(cnApi, disposable);
        mApi = cyApi;
        mCompositeDisposable = disposable;
    }

    @Override
    public void requestWeather(Context context,
                               Location location, @NonNull RequestWeatherCallback callback) {
        Observable<CaiYunMainlyResult> mainly = mApi.getMainlyWeather(
                String.valueOf(location.getLatitude()),
                String.valueOf(location.getLongitude()),
                location.isCurrentPosition(),
                "weathercn%3A" + location.getCityId(),
                15,
                "weather20151024",
                "zUFJoAR2ZVrDy1vF3D07",
                "V10.0.1.0.OAACNFH",
                "10010002",
                false,
                false,
                "gemini",
                "",
                "zh_cn"
        );
        Observable<CaiYunForecastResult> forecast = mApi.getForecastWeather(
                String.valueOf(location.getLatitude()),
                String.valueOf(location.getLongitude()),
                "zh_cn",
                false,
                "weather20151024",
                "weathercn%3A" + location.getCityId(),
                "zUFJoAR2ZVrDy1vF3D07"
        );

        Observable.zip(mainly, forecast, (mainlyResult, forecastResult) ->
                CaiyunResultConverter.convert(context, location, mainlyResult, forecastResult)
        ).compose(SchedulerTransformer.create())
                .subscribe(new ObserverContainer<>(mCompositeDisposable, new BaseObserver<WeatherResultWrapper>() {
                    @Override
                    public void onSucceed(WeatherResultWrapper wrapper) {
                        if (wrapper.result != null) {
                            location.setWeather(wrapper.result);
                            callback.requestWeatherSuccess(location);
                        } else {
                            callback.requestWeatherFailed(location);
                        }
                    }

                    @Override
                    public void onFailed() {
                        callback.requestWeatherFailed(location);
                    }
                }));
    }

    @Override
    public void cancel() {
        super.cancel();
        mCompositeDisposable.clear();
    }

    @Override
    public ChineseCity.CNWeatherSource getSource() {
        return ChineseCity.CNWeatherSource.CAIYUN;
    }
}

File: app/src/main/java/wangdaye/com/geometricweather/weather/services/CaiYunWeatherService.kt
package wangdaye.com.geometricweather.weather.services

import android.content.Context
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import wangdaye.com.geometricweather.common.basic.models.ChineseCity.CNWeatherSource
import wangdaye.com.geometricweather.common.basic.models.Location
import wangdaye.com.geometricweather.common.basic.models.weather.Weather
import wangdaye.com.geometricweather.weather.apis.CNWeatherApi
import wangdaye.com.geometricweather.weather.apis.CaiYunApi
import wangdaye.com.geometricweather.weather.converters.CaiyunResultConverter
import javax.inject.Inject

/**
 * CaiYun weather service.
 */
class CaiYunWeatherService @Inject constructor(private val api: CaiYunApi,
                                               cnApi: CNWeatherApi) : CNWeatherService(cnApi) {

    override val source: CNWeatherSource
        get() = CNWeatherSource.CAIYUN

    override suspend fun getWeather(context: Context, location: Location): Weather? = coroutineScope {
        val mainly = async {
            api.getMainlyWeather(location.latitude.toString(), location.longitude.toString(),
                    location.isCurrentPosition,
                    "weathercn%3A" + location.cityId,
                    15,
                    "weather20151024",
                    "zUFJoAR2ZVrDy1vF3D07",
                    "V10.0.1.0.OAACNFH",
                    "10010002",
                    false,
                    false,
                    "gemini",
                    "",
                    "zh_cn"
            )
        }
        val forecast = async {
            api.getForecastWeather(location.latitude.toString(), location.longitude.toString(),
                    "zh_cn",
                    false,
                    "weather20151024",
                    "weathercn%3A" + location.cityId,
                    "zUFJoAR2ZVrDy1vF3D07"
            )
        }

        return@coroutineScope CaiyunResultConverter.convert(
                context, location, mainly.await(), forecast.await()
        ).result
    }
}